dali/pipeline/operators/util/external_source.cc

```diff
@@ -17,29 +17,50 @@
 namespace dali {
 
 template<>
-void ExternalSource<CPUBackend>::RunImpl(SampleWorkspace *ws) {
-  // Wrap the output tensor around our data
-  auto &output = ws->Output<CPUBackend>(0);
-  cudaStream_t stream = ws->has_stream() ? ws->stream() : 0;
-  if (data_in_tl_) {
-    DALI_ENFORCE(OperatorBase::batch_size_ == static_cast<int>(tl_data_.ntensor()),
-      "Data list provided to ExternalSource needs to have batch_size length.");
-    output.Copy(tl_data_, ws->data_idx(), stream);
-  } else {
-    DALI_ENFORCE(OperatorBase::batch_size_ == static_cast<int>(t_data_.size()),
-      "Data list provided to ExternalSource needs to have batch_size length.");
-    auto &data = t_data_[ws->data_idx()];
-    output.Copy(data, stream);
+void ExternalSource<CPUBackend>::RunImpl(HostWorkspace *ws) {
+  bool is_tl_data;
+  std::unique_ptr<TensorList<CPUBackend>> tl_data;
+  std::unique_ptr<std::vector<Tensor<CPUBackend>>> t_data;
+  {
+    std::unique_lock<std::mutex> busy_lock(busy_m_);
+    cv_.wait(busy_lock, [&data = data_in_tl_]{return !data.empty();});
+    is_tl_data = data_in_tl_.front();
+    data_in_tl_.pop_front();
+    if (is_tl_data) {
+        DALI_ENFORCE(!tl_data_.IsEmpty(), "ExternalSource is empty. Need to feed data first.");
+        tl_data = tl_data_.PopFront();
+        DALI_ENFORCE(OperatorBase::batch_size_ == static_cast<int>(tl_data->ntensor()),
+        "Data list provided to ExternalSource needs to have batch_size length.");
+    } else {
+        DALI_ENFORCE(!t_data_.IsEmpty(), "ExternalSource is empty. Need to feed data first.");
+        t_data = t_data_.PopFront();
+        DALI_ENFORCE(OperatorBase::batch_size_ == static_cast<int>(t_data->size()),
+        "Data list provided to ExternalSource needs to have batch_size length.");
+    }
   }
-
-  std::lock_guard<std::mutex> l(samples_processed_m_);
-  if (++samples_processed_ >= batch_size_) {
-    samples_processed_ = 0;
-    {
-      std::lock_guard<std::mutex> busy_lock(busy_m_);
-      busy_ = false;
+  auto &thread_pool = ws->GetThreadPool();
+  for (int data_idx = 0; data_idx < batch_size_; ++data_idx) {
+    thread_pool.DoWorkWithID([ws, data_idx, is_tl_data, &tl_data, &t_data](int tid) {
+      SampleWorkspace sample;
+      ws->GetSample(&sample, data_idx, tid);
```

Comment:
Do you need to create the SampleWorkspace here, or can you use one of the HostWorkspace::Output(0, sample_idx), or OutputRef(0) and than ask for a pointer to certain tensor?